File: 0125-valid-palindrome/solution.java
class Solution {
    public boolean isPalindrome(String s) {
        int p = 0;
        int q = s.length() - 1;

        while (p < q) {
            while (p < q && !Character.isLetterOrDigit(s.charAt(p)))
                p++;
            while (p < q && !Character.isLetterOrDigit(s.charAt(q)))
                q--;

            char lowerP = s.charAt(p);
            if (lowerP >= 'A' && lowerP <= 'Z') {
                lowerP = (char) (lowerP - ('A' - 'a'));
            }

            char lowerQ = s.charAt(q);
            if (lowerQ >= 'A' && lowerQ <= 'Z') {
                lowerQ = (char) (lowerQ - ('A' - 'a'));
            }

            if (lowerP != lowerQ)
                return false;

            p++;
            q--;
        }

        return true;
    }
}

File: 0125-valid-palindrome/solution.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        p, q = 0, len(s) - 1

        while p < q:
            if not s[p].isalnum():
                p = p + 1
                continue
            
            if not s[q].isalnum():
                q = q - 1
                continue
            
            if s[p].lower() != s[q].lower():
                return False
            
            p += 1
            q -= 1
    
        return True
